### ISSUE TITLE
Adding A Linux Build Agent

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#3455]
Should we add a Linux build agent to complement the Windows and Mac TeamCity build agents we already have?

The benefits:
- Could improve build times by having the Linux code build in parallel to Windows and Mac.
- We could catch build errors that only occur under Mono on Linux.
- We can run Linux unit tests on a actually Linux machine.

The down sides:
- Another agent to manage and maintain.
- Picking the Linux distro to support.

Any thoughts on this?

**Comment by tomspilman** [comment]
@cra0zy @dellis1972 @KonajuGames @thefiddler   ???

**Comment by cra0zy** [comment]
I think that it would be nice to have a Linux build agent. If you need help managing and maintaining it I can help. For the distro, I recomend using Ubuntu since it's the linux disto most people use.

**Comment by WardBenjamin** [comment]
I for one think this is a fantastic idea. The benefits here seem to outweigh the use of another agent. This would help to catch weird bugs like the screen width not resizing, etc. I think the optimally supported distro would be Debian/Ubuntu/whatever, since a major part of all Linux users use Debian or Debian-derived distros. However, I don't think we should rule out Red Hat-based distros like Fedora or anything until others input. Personally, there shouldn't be a guarantee that all distros will work correctly, such as the distro that I use, Arch, which isn't directly related to either of those roots, but people are very active in reporting bugs on the smaller platforms, so there shouldn't be many problems there.

**Comment by tomspilman** [comment]
> For the distro, I recomend using Ubuntu 

 Seems reasonable to me, but I don't know much about Linux.


>  This would help to catch weird bugs like the screen width not resizing, etc.

Assuming we had a unit test for that... yes it would. But at this point we don't have any tests for verifying window management features.

**Comment by WardBenjamin** [comment]
> Assuming we had a unit test for that... yes it would. But at this point we don't have any tests for verifying window management features.
I think that this is going to be one of the most widely encountered issues for the various Linux platforms rather than the code itself, because of the wide range of desktop environments, etc. However, that doesn't rule out the odd compiler peculiarity on one of the platforms. I think that unit tests should be written for this at some point, as well.

**Comment by dellis1972** [comment]
It will probably help.

http://distrowatch.com/dwres.php?resource=popularity

Looks like Mint and Ubuntu are the hot ones atm. So I guess Ubuntu

**Comment by WardBenjamin** [comment]
Anything that works on Ubuntu SHOULD work on Mint as well, since it is
based on Ubuntu and uses Ubuntu packages as well as some of its own.

**Comment by tomspilman** [comment]
Ok... first step is sourcing a machine to run it on.  I'll take stock of what we have around the office and see what we can pull together.  Might use this as another reason to upgrade the main TeamCity machine at the same time.

**Comment by chaosnhatred** [comment]
I know this would be a pain in the ass, but it might be wise to have 2
agents, one a deb based and one rpm based. In my 10 years as a system admin
I ran across a plethora of subtle differences in both, especially video
drivers.

I'd say ubuntu and centos.
If you guys need help with those I am open.

**Comment by tomspilman** [comment]
> I know this would be a pain in the ass, but it might be wise to have 2 agents

We can have as many agents as we like really and they can be located anywhere in the world as long as they are reliable with at least 99% uptime.

We could then run Linux unit tests on both.

**Comment by hach-que** [comment]
You'll want a range of distributions most likely, not just Ubuntu.  The reason is that Mono ships differently on different distributions, so what might build fine on one, won't build correctly on another.

You'll probably want:

 - Arch Linux
 - Debian (the latest stable version)
 - Ubuntu (the latest stable version)
 - Fedora + CentOS
 - openSUSE

Ideally configure each to use the official Xamarin sources, since I think it's reasonable to tell users that's the version of Mono you should have installed.

**Comment by cra0zy** [comment]
If using xamarin source for testing, there is no reason to use multiple distributions because they will all have the same version of mono... Also I think for now 1 Linux build agent should be more than enough, later on more can be added if needed.

**Comment by hach-que** [comment]
The Xamarin sources are packaged differently between different distros (especially Deb / RPM), because different distributions have different file layouts.

In addition, different distributions ship different versions of the C library, threading library, etc. etc. so you really do have to test under each configuration, even if the Xamarin version of Mono is used.

**Comment by tomspilman** [comment]
> You'll want a range of distributions most likely, not just Ubuntu

As long as someone is willing to run them and it has good uptime we can make it a required build for a successful PR.

**Comment by cra0zy** [comment]
@tomspilman What do you think on using something like travis as a Linux build bot? (and I do mean as just a build bot, I just need some way to hook up some native Linux installers)
